java/ql/lib/semmle/code/java/security/UnsafeAndroidAccess.qll

```diff
@@ -44,57 +42,75 @@ private class CrossOriginUrlResourceSink extends JavaScriptEnabledUrlResourceSin
  */
 private class JavaScriptEnabledUrlResourceSink extends UrlResourceSink {
   JavaScriptEnabledUrlResourceSink() {
-    exists(Variable settings |
-      webViewLoadUrl(this.asExpr(), settings) and
-      isJSEnabled(settings)
+    exists(WebViewRef webview |
+      isJSEnabled(webview.getAnAccess()) and
+      webViewLoadUrl(this.asExpr(), webview.getAnAccess())
     )
   }
 
   override string getSinkType() { result = "user input vulnerable to XSS attacks" }
 }
 
-/**
- * Holds if a `WebViewLoadUrlMethod` method is called with the given `urlArg` on a
- * WebView with settings stored in `settings`.
- */
-private predicate webViewLoadUrl(Expr urlArg, Variable settings) {
-  exists(MethodAccess loadUrl, Variable webview, MethodAccess getSettings |
-    loadUrl.getArgument(0) = urlArg and
-    loadUrl.getMethod() instanceof WebViewLoadUrlMethod and
-    loadUrl.getQualifier() = webview.getAnAccess() and
-    getSettings.getMethod() instanceof WebViewGetSettingsMethod and
-    webview.getAnAccess() = getSettings.getQualifier() and
-    settings.getAnAssignedValue() = getSettings
-  )
-}
+private class WebViewRef extends Element {
+  WebViewRef() {
+    this.(RefType).getASourceSupertype*() instanceof TypeWebView or
+    this.(Variable).getType().(RefType).getASourceSupertype*() instanceof TypeWebView
+  }
 
-/**
- * A method allowing any-local-file and cross-origin access in the WebSettings class.
- */
-private class CrossOriginAccessMethod extends Method {
-  CrossOriginAccessMethod() {
-    this.getDeclaringType() instanceof TypeWebSettings and
-    this.hasName(["setAllowUniversalAccessFromFileURLs", "setAllowFileAccessFromFileURLs"])
+  /** Gets an access to this WebView. */
+  Expr getAnAccess() {
+    exists(ThisAccess t | t.getType() = this and result = t |
+      t.isOwnInstanceAccess() or
+      t.isEnclosingInstanceAccess(this)
+    )
+    or
+    result = this.(Variable).getAnAccess()
   }
 }
 
-/**
- * The `setJavaScriptEnabled` method for the webview.
- */
-private class AllowJavaScriptMethod extends Method {
-  AllowJavaScriptMethod() {
-    this.getDeclaringType() instanceof TypeWebSettings and
-    this.hasName("setJavaScriptEnabled")
-  }
+private Expr getUnderlyingExpr(Expr e) {
+  if e instanceof CastExpr or e instanceof UnaryExpr
+  then
+    result = getUnderlyingExpr(e.(CastExpr).getExpr()) or
+    result = getUnderlyingExpr(e.(UnaryExpr).getExpr())
```

Comment:
Clarify what unary exprs that aren't casts you're expecting?

Reply:
Removed `getUnderlyingExpr` in bcba470 for now.